src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,613 @@
+    {
+        return BytesType.instance.getString(value, accessor);
+    }
+
+    @Override
+    public ByteBuffer fromString(String source) throws MarshalException
+    {
+        try
+        {
+            return ByteBufferUtil.hexToBytes(source);
+        }
+        catch (NumberFormatException e)
+        {
+            throw new MarshalException(String.format("cannot parse '%s' as hex bytes", source), e);
+        }
+    }
+
+    @Override
+    public List<AbstractType<?>> subTypes()
+    {
+        return Collections.singletonList(elementType);
+    }
+
+    @Override
+    public String toJSONString(ByteBuffer buffer, ProtocolVersion protocolVersion)
+    {
+        return toJSONString(buffer, ByteBufferAccessor.instance, protocolVersion);
+    }
+
+    @Override
+    public <V> String toJSONString(V value, ValueAccessor<V> accessor, ProtocolVersion protocolVersion)
+    {
+        StringBuilder sb = new StringBuilder();
+        sb.append('[');
+        List<V> split = split(value, accessor);
+        for (int i = 0; i < dimension; i++)
+        {
+            if (i > 0)
+                sb.append(", ");
+            sb.append(elementType.toJSONString(split.get(i), accessor, protocolVersion));
+        }
+        sb.append(']');
+        return sb.toString();
+    }
+
+    @Override
+    public Term fromJSONObject(Object parsed) throws MarshalException
+    {
+        if (parsed instanceof String)
+            parsed = JsonUtils.decodeJson((String) parsed);
+
+        if (!(parsed instanceof List))
+            throw new MarshalException(String.format(
+            "Expected a list, but got a %s: %s", parsed.getClass().getSimpleName(), parsed));
+
+        List<?> list = (List<?>) parsed;
+        if (list.size() != dimension)
+            throw new MarshalException(String.format("List had incorrect size: expected %d but given %d; %s", dimension, list.size(), list));
+        List<Term> terms = new ArrayList<>(list.size());
+        for (Object element : list)
+        {
+            if (element == null)
+                throw new MarshalException("Invalid null element in list");
+            terms.add(elementType.fromJSONObject(element));
+        }
+
+        return new Vectors.DelayedValue<>(this, terms);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        VectorType<?> that = (VectorType<?>) o;
+        return dimension == that.dimension && Objects.equals(elementType, that.elementType);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(elementType, dimension);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toString(false);
+    }
+
+    @Override
+    public String toString(boolean ignoreFreezing)
+    {
+        return getClass().getName() + TypeParser.stringifyVectorParameters(elementType, ignoreFreezing, dimension);
```

Comment:
I think we don't need a separate method on `TypeParser` when this is the only caller:
```suggestion
        return '(' + elementType.toString(ignoreFreezing) + ", " + dimension + ')';
```

Reply:
I had this originally but figured it would be best to have the read/write methods in one place... I like symmetry

Reply:
Yep, it might make sense to keep all the stringify* methods together.